client/examples/repl/main.rs

```diff
@@ -0,0 +1,212 @@
+            )));
+        }
+
+        Ok(())
+    }
+
+    /// Returns the name of the command as String
+    fn name(&self) -> String;
+
+    /// Returns the number of required parameters
+    fn required_parameters(&self) -> Vec<String> {
+        Default::default()
+    }
+
+    /// Returns the error message for the command
+    fn error_message(&self) -> String {
+        let plural = |a: usize| match a {
+            1 => "",
+            _ => "s",
+        };
+        let req = self.required_parameters();
+        format!(
+            "requires {} argument{}: {}",
+            req.len(),
+            plural(req.len()),
+            req.join(" ")
+        )
+    }
+}
+
+#[derive(Default, Clone)]
+pub struct State {
+    stronghold: Rc<RefCell<Option<Stronghold>>>,
+    client_path: Rc<RefCell<Vec<u8>>>,
+}
+
+#[derive(Default)]
+struct Repl {
+    state: State,
+    bindings: HashMap<String, Box<dyn Command>>,
+}
+
+impl Repl {
+    /// Creates a command binding with the command as String and an implementation
+    /// of [`Command`]
+    pub fn bind<C>(&mut self) -> Result<(), ReplError<String>>
+    where
+        C: Command + Default + 'static,
+    {
+        let command: C = Default::default();
+
+        if self.bindings.contains_key(&command.name()) {
+            return Err(ReplError::Invalid(format!(
+                "Command binding already present: '{}'",
+                command.name()
+            )));
+        }
+        self.bindings.insert(command.name(), Box::new(command));
+
+        Ok(())
+    }
+
+    /// Evaluates a command
+    pub fn eval(&self) -> Result<TermAction, ReplError<String>> {
+        match readline() {
+            Ok(line) => {
+                let tk = Tokenizer::try_from(line)?;
+
+                let command = tk.command();
+                let bindings = &self.bindings;
+
+                match bindings.get(command).ok_or_else(|| ReplError::Unknown(command.clone())) {
+                    Ok(f) => {
+                        f.validate(tk.parameter())?;
+                        f.eval(self.state.clone(), tk.parameter())
+                    }
+                    Err(e) => Err(e),
+                }
+            }
+            Err(e) => Err(e),
+        }
+    }
+
+    /// Start the repl
+    pub fn run(&self) {
+        println!("{}", HELP_MESSAGE);
+
+        loop {
+            prompt(Some(self.check_initialized()));
+            match self.eval() {
+                Ok(TermAction::Quit) => break,
+                Ok(TermAction::OkMessage(msg)) => {
+                    println!("{}", msg);
+                }
+                Err(e) => {
+                    eprintln!("{:?}", e);
+                }
+                _ => {}
+            }
+        }
+    }
+
+    /// Checks, if the [`REPL`] state has been initialized
+    fn check_initialized(&self) -> String {
+        let path = self.state.client_path.borrow();
+        match path.is_empty() {
+            true => "[uninitialized]".to_string(),
```

Comment:
The repl panics when calling any command other than `init` after startup, in case you want to fix that.

Reply:
ah, yes. good catch. you're referring to a command being used correctly. fixing that.